cmd/clusterctl/client/config/imagemeta_client.go

```diff
@@ -0,0 +1,136 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/internal/util"
+)
+
+const (
+	imagesConfigKey        = "images"
+	allImageConfig         = "all"
+	CertManagerImageConfig = "cert-manager"
+)
+
+// ImageMetaClient has methods to work with image meta configurations.
+type ImageMetaClient interface {
+	// AlterImage alters an image name according to the current image override configurations.
+	AlterImage(component, image string) (string, error)
+}
+
+// imageMetaClient implements ImageMetaClient.
+type imageMetaClient struct {
+	reader         Reader
+	imageMetaCache map[string]*imageMeta
+}
+
+// ensure imageMetaClient implements ImageMetaClient.
+var _ ImageMetaClient = &imageMetaClient{}
+
+func newImageMetaClient(reader Reader) *imageMetaClient {
+	return &imageMetaClient{
+		reader:         reader,
+		imageMetaCache: map[string]*imageMeta{},
+	}
+}
+
+func (p *imageMetaClient) AlterImage(component, image string) (string, error) {
+	// Gets the image meta that applies to the selected component; if none, returns early
+	meta, err := p.getImageMetaByComponent(component)
+	if err != nil {
+		return "", err
+	}
+	if meta == nil {
+		return image, nil
+	}
+
+	// Apply the image meta to image name
+	return meta.ApplyToImage(image)
+}
+
+// getImageMetaByComponent returns the image meta that applies to the selected component
+func (p *imageMetaClient) getImageMetaByComponent(component string) (*imageMeta, error) {
+	// if the image meta for the component is already known, return it
+	if im, ok := p.imageMetaCache[component]; ok {
+		return im, nil
+	}
+
+	// Otherwise read the image override configurations.
+	var meta map[string]imageMeta
+	if err := p.reader.UnmarshalKey(imagesConfigKey, &meta); err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal image override configurations")
+	}
+
+	// If there are not image override configurations, return.
+	if meta == nil {
+		p.imageMetaCache[component] = nil
+		return nil, nil
+	}
+
+	// Gets the image configuration and to the specific component, and returns the union of the two.
+	m := &imageMeta{}
+	if allMeta, ok := meta[allImageConfig]; ok {
+		m.Union(allMeta)
+	}
+	if componentMeta, ok := meta[component]; ok {
+		m.Union(componentMeta)
+	}
+
+	p.imageMetaCache[component] = m
+	return m, nil
+}
+
+// imageMeta allows to define transformations to apply to the image contained in the YAML manifests.
+type imageMeta struct {
+	// repository sets the container registry to pull images from.
+	Repository string `json:"repository,omitempty"`
+
+	// Tag allows to specify a tag for the images.
+	Tag string `json:"tag,omitempty"`
```

Comment:
custom tags create the upgrade problem, but hopefully nobody will need a custom tag for certmanager.
...or maybe they can remain on an old cert manager tag and update less frequently.

Reply:
yes, this is a known problem, but for the first iteration this is acceptable

Reply:
@neolit123 could you elaborate on what you mean here please?

Reply:
I guess it is referring to the fact that in kubeadm when you pin to an image tag, then the component does not upgrade anymore as usual....

Reply:
^ yes, this relates to the discussion we had yesterday with SUSE folk about etcd mutable upgrades.